lib/hexo/load_config.js

```diff
@@ -4,54 +4,60 @@ const { sep, resolve, join, parse } = require('path');
 const tildify = require('tildify');
 const Theme = require('../theme');
 const Source = require('./source');
-const fs = require('hexo-fs');
+const { exists, readdir } = require('hexo-fs');
 const { magenta } = require('chalk');
 const { deepMerge } = require('hexo-util');
 
-module.exports = ctx => {
+module.exports = async ctx => {
   if (!ctx.env.init) return;
 
   const baseDir = ctx.base_dir;
   let configPath = ctx.config_path;
 
-  return fs.exists(configPath).then(exist => {
-    return exist ? configPath : findConfigPath(configPath);
-  }).then(path => {
-    if (!path) return;
+  const path = await exists(configPath) ? configPath : await findConfigPath(configPath);
+  if (!path) return;
+  configPath = path;
 
-    configPath = path;
-    return ctx.render.render({path});
-  }).then(config => {
-    if (!config || typeof config !== 'object') return;
+  let config = await ctx.render.render({ path });
+  if (!config || typeof config !== 'object') return;
 
-    ctx.log.debug('Config loaded: %s', magenta(tildify(configPath)));
+  ctx.log.debug('Config loaded: %s', magenta(tildify(configPath)));
 
-    ctx.config = deepMerge(ctx.config, config);
-    config = ctx.config;
+  ctx.config = deepMerge(ctx.config, config);
+  config = ctx.config;
 
-    ctx.config_path = configPath;
+  ctx.config_path = configPath;
 
-    config.root = config.root.replace(/\/*$/, '/');
-    config.url = config.url.replace(/\/+$/, '');
+  config.root = config.root.replace(/\/*$/, '/');
+  config.url = config.url.replace(/\/+$/, '');
 
-    ctx.public_dir = resolve(baseDir, config.public_dir) + sep;
-    ctx.source_dir = resolve(baseDir, config.source_dir) + sep;
-    ctx.source = new Source(ctx);
+  ctx.public_dir = resolve(baseDir, config.public_dir) + sep;
+  ctx.source_dir = resolve(baseDir, config.source_dir) + sep;
+  ctx.source = new Source(ctx);
 
-    if (!config.theme) return;
+  if (!config.theme) return;
+
+  const theme = config.theme.toString();
+  config.theme = theme;
+
+  const themeDirFromThemes = join(baseDir, 'themes', theme) + sep; // base_dir/themes/[config.theme]/
+  const themeDirFromNodeModules = join(ctx.plugin_dir, 'hexo-theme-' + theme) + sep; // base_dir/node_modules/hexo-theme-[config.theme]/
+
+  // themeDirFromThemes has higher priority than themeDirFromNodeModules
+  if (await exists(themeDirFromThemes)) {
+    ctx.theme_dir = themeDirFromThemes;
+  } else if (await exists(themeDirFromNodeModules)) {
+    ctx.theme_dir = themeDirFromNodeModules;
+  }
+  ctx.theme_script_dir = join(ctx.theme_dir, 'scripts') + sep;
+  ctx.theme = new Theme(ctx);
```

Comment:
```suggestion
  ctx.theme = new Theme(ctx);
  ctx.theme.ignore = ['**/node_modules/hexo-theme-*/node_modules/**', '**/node_modules/hexo-theme-*/.git/**']
```

see: https://github.com/hexojs/hexo/blob/master/lib/box/index.js#L11

Reply:
It is used in `_readDir`
https://github.com/hexojs/hexo/blob/master/lib/box/index.js#L88

Reply:
Another problem is that whether it is ignored or not, it will affect the performance of watch.

My theme tried to do the same thing in index.js https://github.com/jiangtj/hexo-theme-cake/blob/master/index.js, but when I debug using yarn command (yarn link), Because there are more develop dependencies, `hexo cl && hexo s` will take longer

May be better solution is to listen only to the specified directory, like `source` `scripts` `layout` `languages`

Reply:
This is more symmetrical
```js
  let ignoreCfg;
  if (await exists(themeDirFromThemes)) {
    ctx.theme_dir = themeDirFromThemes;
    ignoreCfg = ['**/themes/*/node_modules/**', '**/themes/*/.git/**'];
  } else if (await exists(themeDirFromNodeModules)) {
    ctx.theme_dir = themeDirFromNodeModules;
    ignoreCfg = ['**/node_modules/hexo-theme-*/node_modules/**', '**/node_modules/hexo-theme-*/.git/**'];
  }
  ctx.theme_script_dir = join(ctx.theme_dir, 'scripts') + sep;
  ctx.theme = new Theme(ctx);
```

And https://github.com/hexojs/hexo/blob/master/lib/box/index.js#L11 can be removed

Reply:
@jiangtj @stevenjoezhang This could be done in another PR then.

Reply:
I agree that another PR can be used to deal with the efficiency issues caused by `watch()`. After resolving the conflict, this PR can be merged

Reply:
@stevenjoezhang Rebased.